osf/models/mixins.py

```diff
@@ -1711,3 +2000,71 @@ def flag_spam(self):
+        :param Object resource: Primary resource where you want to copy attributes
+        :param Object alternative_resource: Backup resource for copying attributes
+        """
+        resource_field_value = getattr(resource, fieldname, None)
+        alternative_resource_field_value = getattr(resource, fieldname, None)
+
+        if resource_field_value:
+            setattr(self, fieldname, resource_field_value)
+        elif alternative_resource_field_value:
+            setattr(self, fieldname, alternative_resource_field_value)
+
+    def stage_m2m_values(self, fieldname, resource, alternative_resource=None):
+        """
+        :param str fieldname: Attribute on model
+        :param Object resource: Primary resource where you want to copy attributes
+        :param Object alternative_resource: Backup resource for copying attributes
+        """
+        resource_field_value = getattr(resource, fieldname, None)
+        alternative_field_value = getattr(alternative_resource, fieldname, None)
+
+        if resource_field_value:
+            return resource_field_value.values_list('pk', flat=True)
+        elif alternative_field_value:
+            return alternative_field_value.values_list('pk', flat=True)
+        else:
+            return []
+
+    def copy_editable_fields(self, resource, auth=None, alternative_resource=None, save=True):
+        """
+        Copy various editable fields from the 'resource' object to the current object.
+        Includes, title, description, category, contributors, node_license, tags, subjects, and affiliated_institutions
+
+        The field on the resource will always supersede the field on the anonymous_resource. For example,
+        copying fields from the draft_registration to the registration.  resource will be a DraftRegistration object,
+        but the alternative_resource will be a Node.  DraftRegistration fields will trump Node fields.
+
+        TODO, add optional logging parameter
+        """
+        self.set_editable_attribute('title', resource, alternative_resource)
+        self.set_editable_attribute('description', resource, alternative_resource)
+        self.set_editable_attribute('category', resource, alternative_resource)
+        self.set_editable_attribute('node_license', resource, alternative_resource)
+
+        # Contributors will always come from "resource", as contributor constraints
+        # will require contributors to be present on the resource
+        self.copy_contributors_from(resource)
+        # Copy unclaimed records for unregistered users
+        self.copy_unclaimed_records(resource)
+
+        self.tags.add(*self.stage_m2m_values('all_tags', resource, alternative_resource))
+        self.subjects.add(*self.stage_m2m_values('subjects', resource, alternative_resource))
+        self.affiliated_institutions.add(*self.stage_m2m_values('affiliated_institutions', resource, alternative_resource))
+
+        if save:
+            self.save()
+
+        return
```

Comment:
A lot of editable fields need to be copied from resource to resource at different stages in the registration process.

1) When creating a draft registration. Copying everything from the node to the draft registration.
2)  when registering a draft node, copying everything from the draft registration back to the draft node (after it's been converted to a regular node)
3) when creating a registration, copying fields from the node to the registration, then overriding with any fields on the draft registration
4) preserving old workflows. When registering node through old API endpoints, copying original node's editable metadata to the draft before registration since draft registrations trump node fields.

Need to do more thinking about logging here. Perhaps an optional logging parameter.